### DISCUSSION TITLE
Importance of local fallback

### DISCUSSION THREAD

**Discussion by AtmanActive** [apohl79/audiogridder#406] [Ideas]
Hi,
Congrats on building this awesome software. 
I'm blown away and will be a loyal backer and supporter going forward.
Maybe I'll even be able to find some time for some pull requests as well.

My primary DAW is Reaper and, once upon a time, I was using Reamote for this, Reaper's obscure remote-FX solution that seems to be dead right now. In almost every aspect, Audiogridder is already ahead and looking awesome. However, there is one aspect where Reamote really shine: transparency.

What I am talking about is the ability of the DAW to just skip Reamote (Audiogridder in this case) processing if a remote server is not found.
Right now, once I add Audiogridder into my project, it stays there forever and just fails silently if there is no working server found. But, this hurts project portability and forces a kind of a location-lock-in on a project.

I understand your build architecture and fully understand that what I am talking about is quite hard to implement in the current state.

But, in my view, something like this is absolutely essential.

Maybe one of the options would be a suggested local-server instance. For example, if the plug-in fails to detect any servers AND there is some data it has to work on (some config within the project), then, a pop-up is shown that Audiogridder can't process audio since there is no server present, but end-user can easily run one on their local machine by following this-and-this page.

Or, as a full-blown solution, having the ability for Audiogridder to load the plug-ins locally itself, like a VST Chain plug-in and run them from there, if no server is found. Yes, yes, I understand, this would be like building a completely different functionality from scratch, but, I believe, in the end, it would be worth it.

To be fully usable in each and every project, Audiogridder needs to add CPU resources when it is able to, but also, get out of the way when it is not able to. 

Please let me know your thoughts.

Thanks.

Cheers!

**Comment by apohl79** [comment]
Thanks for the feedback! Also pull requests are much appreciated!

I think I understand your scenario. In my thinking there is at least two parts to this. First, better local support is needed. I'm still not sure what route I will take, but the easiest way seems to be a better window handling of the server (freely movable, multiple windows, and positioning them beside an AG plugin insert rack). The problem will still be, that the plugin windows will not be created by the DAW process but the server (so cubase for example would just hide all plugin windows, when another processes window gets focused). The alternative is, as you suggested, to create a host plugin - tbh, I don't like that idea - or secondly, make the server manage separate UI sessions automatically and continue to use the UI streaming. But this will definitely cause issues with Waves on macOS (and maybe others), as I have played with this a little bit. So improved window handling will probably be the short term approach.

Secondly - regarding project portability, that I have stumbled upon myself - it is in some way limiting that the remote server is stored in the project session as part of each AG plugin state without being able to change it. I think a convenient way to switch remote servers in a project is needed, across all plugins (at least all plugins that use the same server). I like the idea of a popup. Though I have not finished my thinking here, but I'm tending towards introducing a tray application for the AG plugins as well, move the settings there and add functions like this to it. A better channel view is needed too, as the DAW insert views are a bit limiting. But thats another story. Tha means if you replace a DSP server, you should be able to point all plugins to a new one with a "simple" click basically. If thats possible, switching to a local server will be easy too or adding this functionality to a popup.

**Comment by AtmanActive** [comment]
There is one in-between development step that could come handy in these situations: Load/Save FX Chain Data.
I mean, sooner or later Audiogridder will have to have an ability to save it's chain data and load it back. Upon loading, there could be an option to use some other server or even a local one.
In this scenario, it could help end-users to migrate to other servers or local processing.

The worst thing to do is to lose end-user's data (ouch). :-)

**Reply by apohl79** [reply]
You are right, and it's on the list already.

**Reply by Fittkenny** [reply]
I can’t wait!
